state/executor.go

```diff
@@ -303,6 +307,33 @@ func (e *BlockExecutor) updateState(state types.State, block *types.Block, final
 		state.ConsensusParams = nextParamsProto
 	}
 
+	nValSet := state.NextValidators.Copy()
+	lastHeightValSetChanged := state.LastHeightValidatorsChanged
+
+	if len(nValSet.Validators) > 0 {
+		if len(validatorUpdates) > 0 {
+			err := nValSet.UpdateWithChangeSet(validatorUpdates)
+			if err != nil {
+				if err.Error() != ErrEmptyValSetGenerated.Error() {
+					return state, err
+				}
+				nValSet = &cmtypes.ValidatorSet{
+					Validators: make([]*cmtypes.Validator, 0),
+					Proposer:   nil,
+				}
+			}
+			// Change results from this height but only applies to the next next height.
+			lastHeightValSetChanged = int64(block.SignedHeader.Header.Height() + 1 + 1)
+		}
+
+		if len(nValSet.Validators) > 0 {
+			nValSet.IncrementProposerPriority(1)
+		}
+		// TODO(tzdybal):  right now, it's for backward compatibility, may need to change this
+	} else if len(validatorUpdates) > 0 {
+		return state, ErrAddingValidatorToBased
+	}
```

Comment:
You should check for error conditions first so you can quite early and reduce the number of if conditions needed. 

```suggestion
	// TODO(tzdybal):  right now, it's for backward compatibility, may need to change this
	if len(validatorUpdates) > 0 {
		return state, ErrAddingValidatorToBased
	}
	if len(nValSet.Validators) > 0 {
			err := nValSet.UpdateWithChangeSet(validatorUpdates)
			if err != nil {
				if err.Error() != ErrEmptyValSetGenerated.Error() {
					return state, err
				}
				nValSet = &cmtypes.ValidatorSet{
					Validators: make([]*cmtypes.Validator, 0),
					Proposer:   nil,
				}
			}
			// Change results from this height but only applies to the next next height.
			lastHeightValSetChanged = int64(block.SignedHeader.Header.Height() + 1 + 1)

		if len(nValSet.Validators) > 0 {
			nValSet.IncrementProposerPriority(1)
		}
	}
```